File: tools/check_backend_metadata.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Validate each sibling backend's .quixicore/backend.yaml against
# registry/backend-metadata.schema.yaml.
#
# The schema is JSON-Schema-shaped but nothing had ever run it; this is a
# hand-rolled validator for exactly the constructs the schema uses (required,
# enum, pattern, const, additional_properties, typed arrays). Runs locally —
# CI cannot see the sibling checkouts.
#
# Usage: ruby tools/check_backend_metadata.rb   # exit 1 on any violation

require "yaml"

ROOT = File.expand_path("..", __dir__)
SCHEMA = YAML.load_file(File.join(ROOT, "registry", "backend-metadata.schema.yaml")).fetch("schema")

BACKEND_DIRS = %w[
  QuixiCore-Metal QuixiCore-CUDA QuixiCore-ROCm
  QuixiCore-XPU QuixiCore-Gaudi QuixiCore-CPU
].freeze

def validate(doc, schema, path, problems)
  case schema["type"]
  when "object"
    unless doc.is_a?(Hash)
      problems << "#{path}: expected object, got #{doc.class}"
      return
    end
    (schema["required"] || []).each do |key|
      problems << "#{path}: missing required key '#{key}'" unless doc.key?(key)
    end
    props = schema["properties"] || {}
    if schema["additional_properties"] == false
      (doc.keys - props.keys).each do |extra|
        problems << "#{path}: unknown key '#{extra}'"
      end
    end
    doc.each do |key, value|
      validate(value, props[key], "#{path}.#{key}", problems) if props[key]
    end
  when "string"
    unless doc.is_a?(String)
      problems << "#{path}: expected string, got #{doc.class}"
      return
    end
    if (enum = schema["enum"]) && !enum.include?(doc)
      problems << "#{path}: '#{doc}' not in #{enum.inspect}"
    end
    if (pattern = schema["pattern"]) && doc !~ Regexp.new(pattern)
      problems << "#{path}: '#{doc}' does not match #{pattern}"
    end
    if (const = schema["const"]) && doc != const
      problems << "#{path}: '#{doc}' != required constant '#{const}'"
    end
  when "array"
    unless doc.is_a?(Array)
      problems << "#{path}: expected array, got #{doc.class}"
      return
    end
    if (min = schema["min_items"]) && doc.size < min
      problems << "#{path}: needs at least #{min} item(s)"
    end
    doc.each_with_index do |item, i|
      validate(item, schema["items"], "#{path}[#{i}]", problems) if schema["items"]
    end
  end
end

failures = 0
BACKEND_DIRS.each do |dir|
  file = File.join(ROOT, dir, ".quixicore", "backend.yaml")
  unless File.exist?(file)
    puts "#{dir}: MISSING .quixicore/backend.yaml"
    failures += 1
    next
  end
  problems = []
  validate(YAML.load_file(file), SCHEMA, dir, problems)
  if problems.any?
    failures += 1
    puts "#{dir}: FAIL"
    problems.each { |p| puts "  #{p}" }
  else
    puts "#{dir}: ok"
  end
end

puts "#{failures.zero? ? 'OK' : 'FAIL'}: backend metadata schema"
exit(failures.zero? ? 0 : 1)

File: tools/check_backend_metadata.py
#!/usr/bin/env python3
"""Validate each sibling backend's .quixicore/backend.yaml against
registry/backend-metadata.schema.yaml.

The schema is JSON-Schema-shaped; this is a hand-rolled validator for exactly
the constructs it uses (required, enum, pattern, const, additional_properties,
typed arrays). Runs locally — CI cannot see the sibling checkouts. Stdlib
only; YAML via tools/qc_yaml.py.

Usage: python3 tools/check_backend_metadata.py   # exit 1 on any violation
"""

import os
import re
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import qc_yaml  # noqa: E402

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SCHEMA = qc_yaml.load_file(
    os.path.join(ROOT, "registry", "backend-metadata.schema.yaml")
)["schema"]

BACKEND_DIRS = [
    "QuixiCore-Metal", "QuixiCore-CUDA", "QuixiCore-ROCm",
    "QuixiCore-XPU", "QuixiCore-Gaudi", "QuixiCore-CPU",
]


def validate(doc, schema, path, problems):
    stype = schema.get("type")
    if stype == "object":
        if not isinstance(doc, dict):
            problems.append(f"{path}: expected object, got {type(doc).__name__}")
            return
        for key in schema.get("required") or []:
            if key not in doc:
                problems.append(f"{path}: missing required key '{key}'")
        props = schema.get("properties") or {}
        if schema.get("additional_properties") is False:
            for extra in doc:
                if extra not in props:
                    problems.append(f"{path}: unknown key '{extra}'")
        for key, value in doc.items():
            if key in props:
                validate(value, props[key], f"{path}.{key}", problems)
    elif stype == "string":
        if not isinstance(doc, str):
            problems.append(f"{path}: expected string, got {type(doc).__name__}")
            return
        enum = schema.get("enum")
        if enum and doc not in enum:
            problems.append(f"{path}: '{doc}' not in {enum!r}")
        pattern = schema.get("pattern")
        if pattern and not re.search(pattern, doc):
            problems.append(f"{path}: '{doc}' does not match {pattern}")
        const = schema.get("const")
        if const and doc != const:
            problems.append(f"{path}: '{doc}' != required constant '{const}'")
    elif stype == "array":
        if not isinstance(doc, list):
            problems.append(f"{path}: expected array, got {type(doc).__name__}")
            return
        min_items = schema.get("min_items")
        if min_items and len(doc) < min_items:
            problems.append(f"{path}: needs at least {min_items} item(s)")
        items = schema.get("items")
        if items:
            for i, item in enumerate(doc):
                validate(item, items, f"{path}[{i}]", problems)


def main():
    failures = 0
    for dirname in BACKEND_DIRS:
        file = os.path.join(ROOT, dirname, ".quixicore", "backend.yaml")
        if not os.path.exists(file):
            print(f"{dirname}: MISSING .quixicore/backend.yaml")
            failures += 1
            continue
        problems = []
        validate(qc_yaml.load_file(file), SCHEMA, dirname, problems)
        if problems:
            failures += 1
            print(f"{dirname}: FAIL")
            for p in problems:
                print(f"  {p}")
        else:
            print(f"{dirname}: ok")
    print(f"{'OK' if not failures else 'FAIL'}: backend metadata schema")
    return 0 if not failures else 1


if __name__ == "__main__":
    sys.exit(main())
